### ISSUE TITLE
GetAllControls should be banned in a core code

### ISSUE THREAD

**Issue by olegabr** [qcubed/qcubed#883]
Now there are many calls to this method in the very core Run method and other methods it calls directly or indirectly. it is a problem because it means that the complexity of qcubed AJAX call is O(N), where N is a number of controls on a form. It is bad on it's own. It limits the complexity of qcubed-based applications. And it violates the very important principle that "you should not pay for what you don't use"

The solution is to use some sort of listener or events pattern. For example, this one: http://php.net/manual/en/intro.event.php

**Comment by olegabr** [comment]
Same for the `IsCheckableControlRendered` method. There is a code block in the Run method which is needed only to support it. The `IsCheckableControlRendered` method is used in the `ParsePostData` methos of a few chackable controls. It looks like a hack, not an architecture.

**Comment by spekary** [comment]
Agreed, if possible. I tried to reduce the number of those, but was not able to eliminate it.

FYI, all uses of ```IsCheckableControlRendered``` happen only in non-ajax code. 

```IsCheckableControlRendered``` could use empty() instead of array_key_exists for very small speed boost.

**Comment by spekary** [comment]
Taking another look at this. ```IsCheckableControlRendered``` is no longer used. I recently made changes that makes more use of checkable control tracking though.

I have done a few passes through the drawing code. I tried to reduce the number of times GetAllControls is called, but the code is very complex, and we need to call it a few times.

That said, I think it would take a very big form indeed for the speed of this to affect drawing. Probably the amount of HTML being sent to the browser will have a bigger effect than this.
